cinema/views.py

```diff
@@ -46,13 +95,76 @@ def get_serializer_class(self):
 
 class MovieSessionViewSet(viewsets.ModelViewSet):
     queryset = MovieSession.objects.all()
-    serializer_class = MovieSessionSerializer
 
-    def get_serializer_class(self):
+    def get_queryset(self) -> QuerySet:
+        self.queryset = self.queryset.select_related(
+            "movie",
+            "cinema_hall"
+        ).annotate(
+            tickets_available=(
+                F("cinema_hall__rows") * F("cinema_hall__seats_in_row")
+            ) - Count("tickets")
+        )
+
+        date = self.request.query_params.get("date")
+        movie = self.request.query_params.get("movie")
+        if date:
+            try:
+                date = datetime.strptime(date, "%Y-%m-%d").date()
+            except ValueError:
+                return Response(
+                    {"error": "Invalid date format. Please use YYYY-MM-DD."},
+                    status=status.HTTP_400_BAD_REQUEST
+                )
```

Comment:
In the `MovieSessionViewSet`, returning a `Response` object directly from `get_queryset` is not standard practice. Instead, consider raising a `serializers.ValidationError` or handling the error in a different method to maintain consistency with the rest of the viewset.